File: decisions/assets/ADR-003/examples/python/main.py
#!/usr/bin/env python3

import jwt
import os
import requests
import time


def main():

    gh_app_id = os.environ.get("GITHUB_APP_ID")
    gh_app_pk_file = os.environ.get("GITHUB_APP_PK_FILE")
    gh_org = os.environ.get("GITHUB_ORG")

    if not gh_app_id or not gh_app_pk_file or not gh_org:
        raise ValueError("Environment variables GITHUB_APP_ID, GITHUB_APP_PK_FILE and GITHUB_ORG must be passed to this program.")

    jwt_token = get_jwt_token(gh_app_id, gh_app_pk_file)
    installation_id = get_installation_id(jwt_token, gh_org)
    access_token = get_access_token(jwt_token, installation_id)

    print(f"GITHUB_TOKEN={access_token}")


def get_jwt_token(gh_app_id, gh_app_pk_file):

    with open(gh_app_pk_file, "rb") as file:
        private_key = file.read()
    payload = {"iat": int(time.time()), "exp": int(time.time()) + 600, "iss": gh_app_id}
    jwt_token = jwt.encode(payload, private_key, algorithm="RS256")

    return jwt_token


def get_installation_id(jwt_token, gh_org):

    headers = {
        "Authorization": f"Bearer {jwt_token}",
        "Accept": "application/vnd.github.v3+json",
    }
    url = "https://api.github.com/app/installations"
    response = requests.get(url, headers=headers)

    installation_id = None
    for installation in response.json():
        if installation["account"]["login"] == gh_org:
            installation_id = installation["id"]
            break

    return installation_id


def get_access_token(jwt_token, installation_id):

    headers = {
        "Authorization": f"Bearer {jwt_token}",
        "Accept": "application/vnd.github.v3+json",
    }
    url = f"https://api.github.com/app/installations/{installation_id}/access_tokens"
    response = requests.post(url, headers=headers)

    return response.json().get("token")


if __name__ == "__main__":
    main()

File: decisions/assets/ADR-003/examples/golang/main.go
package main

import (
	"crypto/x509"
	"encoding/json"
	"encoding/pem"
	"fmt"
	"io/ioutil"
	"log"
	"os"
	"time"

	"github.com/go-resty/resty/v2"
	"github.com/golang-jwt/jwt"
)

type Installation struct {
	ID      int `json:"id"`
	Account struct {
		Login string `json:"login"`
	} `json:"account"`
}

func main() {

	ghAppId := os.Getenv("GITHUB_APP_ID")
	ghAppPkFile := os.Getenv("GITHUB_APP_PK_FILE")
	ghOrg := os.Getenv("GITHUB_ORG")

	if ghAppId == "" || ghAppPkFile == "" || ghOrg == "" {
		log.Fatalf("Environment variables GITHUB_APP_ID, GITHUB_APP_PK_FILE and GITHUB_ORG must be passed to this program.")
	}

	jwtToken := getJwtToken(ghAppId, ghAppPkFile)
	installationId := getInstallationId(jwtToken, ghOrg)
	accessToken := getAccessToken(jwtToken, installationId)

	fmt.Printf("GITHUB_TOKEN=%s\n", accessToken)
}

func getJwtToken(ghAppId string, ghAppPkFile string) string {

	pemContent, _ := ioutil.ReadFile(ghAppPkFile)
	block, _ := pem.Decode(pemContent)
	privateKey, _ := x509.ParsePKCS1PrivateKey(block.Bytes)
	token := jwt.NewWithClaims(jwt.SigningMethodRS256, jwt.MapClaims{
		"iat": time.Now().Unix(),
		"exp": time.Now().Add(10 * time.Minute).Unix(),
		"iss": ghAppId,
	})
	jwtToken, _ := token.SignedString(privateKey)

	return jwtToken
}

func getInstallationId(jwtToken string, ghOrg string) int {

	client := resty.New()
	resp, _ := client.R().
		SetHeader("Authorization", "Bearer "+jwtToken).
		SetHeader("Accept", "application/vnd.github.v3+json").
		Get("https://api.github.com/app/installations")

	var installations []Installation
	json.Unmarshal(resp.Body(), &installations)
	installationId := 0
	for _, installation := range installations {
		if installation.Account.Login == ghOrg {
			installationId = installation.ID
		}
	}

	return installationId
}

func getAccessToken(jwtToken string, installationId int) string {

	client := resty.New()
	resp, _ := client.R().
		SetHeader("Authorization", "Bearer "+jwtToken).
		SetHeader("Accept", "application/vnd.github.v3+json").
		Post(fmt.Sprintf("https://api.github.com/app/installations/%d/access_tokens", installationId))

	var result map[string]interface{}
	json.Unmarshal(resp.Body(), &result)

	return result["token"].(string)
}
